src/uu/head/src/head.rs

```diff
@@ -378,30 +379,51 @@ where
 
     let mut buffer = [0u8; BUF_SIZE];
 
-    let mut i = 0u64;
     let mut lines = 0u64;
+    let mut check_last_byte_first_loop = true;
+    let mut bytes_remaining_to_search = file_size;
 
     loop {
         // the casts here are ok, `buffer.len()` should never be above a few k
-        let bytes_remaining_to_search = file_size - i;
-        let bytes_to_read_this_loop = bytes_remaining_to_search.min(BUF_SIZE.try_into().unwrap());
+        let bytes_to_read_this_loop =
+            bytes_remaining_to_search.min(buffer.len().try_into().unwrap());
         let read_start_offset = bytes_remaining_to_search - bytes_to_read_this_loop;
         let buffer = &mut buffer[..bytes_to_read_this_loop.try_into().unwrap()];
+        bytes_remaining_to_search -= bytes_to_read_this_loop;
 
         input.seek(SeekFrom::Start(read_start_offset))?;
         input.read_exact(buffer)?;
-        for byte in buffer.iter().rev() {
-            if byte == &separator {
-                lines += 1;
-            }
-            // if it were just `n`,
+
+        // Unfortunately need special handling for the case that the input file doesn't have
+        // a terminating `separator` character.
+        // If the input file doesn't end with a `separator` character, add an extra line to our
+        // `line` counter. In the case that `n` is 0 we need to return here since we've
+        // obviously found our 0th-line-from-the-end offset.
+        if check_last_byte_first_loop {
+            check_last_byte_first_loop = false;
+            match buffer.last() {
+                Some(last_byte_of_file) if (last_byte_of_file != &separator) => {
+                    if n == 0 {
+                        input.rewind()?;
+                        return Ok(file_size);
+                    }
+                    assert_eq!(lines, 0);
+                    lines = 1;
+                }
+                _ => (),
+            };
```

Comment:
I don't think the match is necessary, something like might be better:
```suggestion
    if let Some(last_byte_of_file) = buffer.last() {
    if last_byte_of_file != &separator {
        if n == 0 {
            input.rewind()?;
            return Ok(file_size);
        }
        assert_eq!(lines, 0);
        lines = 1;
    }
}
```

Reply:
Yah, agree this is cleaner.
Code is updated thanks.

Reply:
@sylvestre - hey Sylvestre, could I get your eyes on this PR again please? It's been sitting for a while but I think it's good to do. Am happy to make further changes if you think that's require.
Karl